clap_complete/src/engine/complete.rs

```diff
@@ -59,6 +59,13 @@ pub fn complete(
         } else if arg.is_escape() {
             is_escaped = true;
         } else if let Some((flag, value)) = arg.to_long() {
+            if let ParseState::Opt((opt, count)) = current_state {
+                if opt.is_allow_hyphen_values_set() {
+                    next_state = parse_opt(opt, count);
+                    continue;
+                }
+            }
+
             if let Ok(flag) = flag {
```

Comment:
Could we collapse down the short and long cases by having
```suggestion
        } else if arg.is_escape() {
            is_escaped = true;
        } else if let Some(ParseState::Opt((opt, count))) = opt.is_allow_hyphen_values_set.then(current_state) {
            next_state = parse_opt(opt, count);
            continue;
        } else if let Some((flag, value)) = arg.to_long() {
            if let Ok(flag) = flag {
```
(a previous commit had something like this but it didn't properly fall through when the parse state was different)

Reply:
Oh, thats circular but something in that direction could be done (e.g. a function like `pos_allows_hyphens`)

Reply:
This seems to do the thing
```rust
...
            is_escaped = true;
        } else if opt_allows_hyphen(&current_state, &arg) {
            match current_state {
                ParseState::Opt((opt, count)) => next_state = parse_opt_value(opt, count),
                _ => unreachable!("else branch is only reachable in Opt state"),
            }
        } else if let Some((flag, value)) = arg.to_long() {
...

fn opt_allows_hyphen(state: &ParseState<'_>, arg: &clap_lex::ParsedArg<'_>) -> bool {
    let val = arg.to_value_os();
    if val.starts_with("-") {
        if let ParseState::Opt((opt, _)) = state {
            return opt.is_allow_hyphen_values_set();
        }
    }

    false
}
```

Reply:
Good idea.  Thanks!